Kernel/Arch/riscv64/Interrupts.cpp

```diff
@@ -31,6 +34,85 @@ void dump_registers(RegisterState const& regs)
     dbgln("s1={:p} s2={:p} s3={:p} s4={:p} s5={:p} s6={:p} s7={:p} s8={:p} s9={:p} s10={:p} s11={:p}", regs.x[8], regs.x[17], regs.x[18], regs.x[19], regs.x[20], regs.x[21], regs.x[22], regs.x[23], regs.x[24], regs.x[25], regs.x[26]);
 }
 
+extern "C" void trap_handler(TrapFrame& trap_frame);
+extern "C" void trap_handler(TrapFrame& trap_frame)
+{
+    auto scause = trap_frame.regs->scause;
+
+    if ((to_underlying(scause) & RISCV64::CSR::SCAUSE_INTERRUPT_MASK) != 0) {
+        // Interrupt
+
+        Processor::current().enter_trap(trap_frame, true);
+
+        auto interrupt_number = to_underlying(scause) & ~RISCV64::CSR::SCAUSE_INTERRUPT_MASK;
+
+        auto* handler = s_interrupt_handlers[interrupt_number];
+        VERIFY(handler);
+        handler->increment_call_count();
+        handler->handle_interrupt(*trap_frame.regs);
+        handler->eoi();
+
+        Processor::current().exit_trap(trap_frame);
+    } else {
+        // Exception
+
+        Processor::current().enter_trap(trap_frame, false);
+        Processor::enable_interrupts();
+
+        using enum RISCV64::CSR::SCAUSE;
+        switch (scause) {
+        case InstructionAddressMisaligned:
+        case LoadAddressMisaligned:
+        case StoreOrAMOAddressMisaligned:
+            handle_crash(*trap_frame.regs, "Unaligned memory access", SIGBUS, false);
+            break;
+
+        case InstructionAccessFault:
+        case LoadAccessFault:
+        case StoreOrAMOAccessFault:
+            handle_crash(*trap_frame.regs, "Memory access fault", SIGBUS, false);
+            break;
+
+        case IllegalInstrction:
+            handle_crash(*trap_frame.regs, "Illegal instruction", SIGILL, false);
+            break;
+
+        case InstructionPageFault:
+        case LoadPageFault:
+        case StoreOrAMOPageFault: {
+            PageFault fault { VirtualAddress(trap_frame.regs->stval) };
+
+            if (scause == InstructionPageFault)
+                fault.set_instruction_fetch(true);
+            else if (scause == LoadPageFault)
+                fault.set_access(PageFault::Access::Read);
+            else if (scause == StoreOrAMOPageFault)
+                fault.set_access(PageFault::Access::Write);
+
+            // FIXME: detect correct type by walking page table or rewrite MM to not depend on processor-provided reason
```

Comment:
Maybe specify that some implementations indeed report page fault reasons incorrectly since the spec allows it in some circumstances (I remember us talking about this at some point...)

Reply:
I think you mean that implementations are allowed to make stval always 0?
But no, this is not what the comment refers to.
RISC-V only tells you the original access type (r/w/x) to the faulting address. It does not say *why* it failed. But MM needs that information, so I forced it to be a protection violation for now.

Reply:
Yes, that's it. Please document that.

Reply:
I also added a note about stval always being 0.
We should also note that Sstvala required when we document the RISC-V port, next to other extension we reuquire/support.

Is the comment ok?